src/IECoreDelight/Renderer.cpp

```diff
@@ -1104,25 +1135,28 @@ class DelightRenderer final : public IECoreScenePreview::Renderer
 					m_oversampling = 9;
 				}
 			}
+			else if(
+				name == g_maxLengthDiffuseOptionName ||
+				name == g_maxLengthHairOptionName ||
+				name == g_maxLengthReflectionOptionName ||
+				name == g_maxLengthRefractionOptionName ||
+				name == g_maxLengthSpecularOptionName ||
+				name == g_maxLengthVolumeOptionName ||
+				name == g_clampIndirectOptionName
+			)
+			{
+				// These parameters fail to set properly if they are not doubles
+				auto data = runTimeCast<const FloatData>( value );
+				setNSIGlobalOption( m_context, name, data ? new DoubleData( data->readable() ) : nullptr );
+			}
+			else if( name == g_showMultipleScatteringOptionName )
+			{
+				auto data = runTimeCast<const BoolData>( value );
+				setNSIGlobalOption( m_context, name, data ? new DoubleData( data->readable() ) : nullptr );
+			}
 			else if( boost::starts_with( name.string(), "dl:" ) )
 			{
-				if( value )
-				{
-					if( const Data *data = reportedCast<const Data>( value, "option", name ) )
-					{
-						ParameterList params;
-						params.add( name.c_str() + 3, data );
-						NSISetAttribute( m_context, NSI_SCENE_GLOBAL, params.size(), params.data() );
-					}
-					else
-					{
-						NSIDeleteAttribute( m_context, NSI_SCENE_GLOBAL, name.c_str() + 3 );
-					}
-				}
-				else
-				{
-					NSIDeleteAttribute( m_context, NSI_SCENE_GLOBAL, name.c_str() + 3 );
-				}
```

Comment:
Straw man proposal for addressing the leak and my other comments : keep this block as it was, remove the new blocks above, and delegate the conversions to a separate function :

```
else if( boost::starts_with( name.string(), "dl:" ) )
{
     ConstDataPtr conformedValue = conformOptionValue();
     value = conformedValue ? conformedValue.get() : value;
     // Process value as before
}
```

Reply:
That works out quite nicely, including getting the `+3` string slice closer to where it makes sense. I did that change and squashed down to https://github.com/GafferHQ/gaffer/pull/5333/commits/0ca905e98abe63931686805e3f9a5d1ae018d175